data-plane/src/server/data_plane_server.rs

```diff
@@ -54,88 +56,353 @@ where
 
     println!("TLS Server Created - Listening for new connections.");
     loop {
-        let stream = match server.accept().await {
+        let mut stream = match server.accept().await {
             Ok(stream) => stream,
             Err(tls_err) => {
                 eprintln!("An error occurred while accepting the incoming connection — {tls_err}");
                 continue;
             }
         };
 
-        let server = http_server.clone();
         let e3_client_for_connection = e3_client.clone();
         let tx_for_connection = tx.clone();
         tokio::spawn(async move {
             let e3_client_for_tcp = e3_client_for_connection.clone();
             let tx_for_tcp = tx_for_connection.clone();
             let remote_ip = stream.get_remote_addr();
-            let sent_response = server
-                .serve_connection(
-                    stream,
-                    service_fn(|mut req: Request<Body>| {
-                        let e3_client_for_req = e3_client_for_tcp.clone();
-                        let feature_context = FeatureContext::get();
-                        let cage_context = CAGE_CONTEXT.get().expect("Couldn't get cage context");
-                        let tx_for_req = tx_for_tcp.clone();
-                        let remote_ip = remote_ip.clone();
-                        async move {
-                            let (mut trx_context, request_timer) = init_trx(cage_context, &req);
-                            let trx_id = trx_context.get_trx_id();
-                            if remote_ip.is_some() {
-                              trx_context.remote_ip(remote_ip.clone());
-                              add_remote_ip_to_forwarded_for_header(&mut req, remote_ip.as_deref().unwrap());
-                            }
 
-                            let trx_logging_enabled = feature_context.trx_logging_enabled;
+            let buffer = match process_stream(&mut stream).await {
+                Ok(stream) => stream,
+                Err(e) => {
+                    eprintln!("Failed to read the request stream — {e:?}");
+                    shutdown_conn(&mut stream).await;
+                    return;
+                }
+            };
 
-                            if  trx_logging_enabled {
-                                add_ev_ctx_header_to_request(&mut req, &trx_id);
+            let mut headers = [httparse::EMPTY_HEADER; 64];
+            let mut req = httparse::Request::new(&mut headers);
+            match req.parse(&buffer) {
+                Ok(Status::Complete(remaining)) => {
+                    let headers = &req.headers;
+                    let body_buffer = &buffer[remaining..];
+                    let is_websocket = headers.iter().any(|header| {
+                        header.name.to_ascii_lowercase() == "upgrade"
+                            && header.value.to_ascii_lowercase() == "websocket".as_bytes()
+                    });
+                    let not_http = req.method.is_none();
```

Comment:
Is it possible to hit the complete arm with Method undefined?

Reply:
Don't think because it wouldn't be http so but added the check anyway because its an option in the library